packages/core/src/transport/sendWithRetryStrategy.ts

```diff
@@ -0,0 +1,154 @@
+import type { Payload, HttpResponse } from './httpRequest'
+
+export const MAX_ONGOING_BYTES_COUNT = 80 * ONE_KIBI_BYTE
+export const MAX_ONGOING_REQUESTS = 32
+export const MAX_QUEUE_BYTES_COUNT = 3 * ONE_MEBI_BYTE
+export const MAX_BACKOFF_TIME = 256 * ONE_SECOND
+export const INITIAL_BACKOFF_TIME = ONE_SECOND
+
+export interface RetryState {
+  isIntakeAvailable: boolean
+  currentBackoffTime: number
+  bandwidthMonitor: ReturnType<typeof newBandwidthMonitor>
+  queuedPayloads: ReturnType<typeof newPayloadQueue>
+}
+
+type SendStrategy = (payload: Payload, onResponse: (r: HttpResponse) => void) => void
+
+export function sendWithRetryStrategy(payload: Payload, state: RetryState, sendStrategy: SendStrategy) {
+  if (state.isIntakeAvailable && state.bandwidthMonitor.canHandle(payload)) {
+    send(payload, state, sendStrategy, {
+      onSuccess: () => sendNextPayload(state, sendStrategy),
+      onFailure: () => {
+        state.queuedPayloads.enqueue(payload)
+        scheduleRetry(state, sendStrategy)
+      },
+    })
+    sendNextPayload(state, sendStrategy)
+  } else {
+    state.queuedPayloads.enqueue(payload)
+  }
+}
+
+function scheduleRetry(state: RetryState, sendStrategy: SendStrategy) {
+  if (state.bandwidthMonitor.ongoingRequestCount !== 0) {
+    // avoid to enter a retry phase if another ongoing request can succeed
+    return
+  }
+  setTimeout(
+    monitor(() => {
+      const payload = state.queuedPayloads.first()
+      send(payload, state, sendStrategy, {
+        onSuccess: () => {
+          state.queuedPayloads.dequeue()
+          addTelemetryDebug('resuming after intake failure', {
+            currentBackoffTime: state.currentBackoffTime,
+            queuedPayloadCount: state.queuedPayloads.size(),
+            queuedPayloadBytesCount: state.queuedPayloads.bytesCount,
+          })
+          state.currentBackoffTime = INITIAL_BACKOFF_TIME
+          sendNextPayload(state, sendStrategy)
+        },
+        onFailure: () => {
+          state.currentBackoffTime = Math.min(MAX_BACKOFF_TIME, state.currentBackoffTime * 2)
+          scheduleRetry(state, sendStrategy)
+        },
+      })
+    }),
+    state.currentBackoffTime
+  )
+}
+
+function sendNextPayload(state: RetryState, sendStrategy: SendStrategy) {
+  const nextPayload = state.queuedPayloads.dequeue()
+  if (nextPayload) {
+    sendWithRetryStrategy(nextPayload, state, sendStrategy)
+  }
+}
+
+function send(
+  payload: Payload,
+  state: RetryState,
+  sendStrategy: SendStrategy,
+  { onSuccess, onFailure }: { onSuccess: () => void; onFailure: () => void }
+) {
+  state.bandwidthMonitor.add(payload)
+  sendStrategy(payload, (response) => {
+    state.bandwidthMonitor.remove(payload)
+    if (wasRequestSuccessful(response)) {
+      state.isIntakeAvailable = true
+      onSuccess()
+    } else {
+      state.isIntakeAvailable = false
+      onFailure()
+    }
+  })
+}
+
+function wasRequestSuccessful(response: HttpResponse) {
+  return response.status < 500
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏retry payloads when internet is down:
```suggestion
  return response.status === 0 || response.status < 500
```

Reply:
Indeed, I've seen network status check listed as improvement and discarded the retry when request fail due to network down :facepal